From 3b687f529fa18d6248cb4088aa8cdcf1503c1872 Mon Sep 17 00:00:00 2001
From: Baozhu Zuo <zuobaozhu@gmail.com>
Date: Thu, 7 Sep 2017 09:22:40 +0800
Subject: [PATCH] include:configs: use respeaker boot first

---
 include/configs/evb_rk3229.h | 66 +++++++++++++++++++++++++++-----------------
 1 file changed, 40 insertions(+), 26 deletions(-)

diff --git a/include/configs/evb_rk3229.h b/include/configs/evb_rk3229.h
index df441f6..d7ba1f2 100644
--- a/include/configs/evb_rk3229.h
+++ b/include/configs/evb_rk3229.h
@@ -14,39 +14,17 @@
 #undef CONFIG_ENV_SIZE
 #define CONFIG_ENV_SIZE                 (32 << 10)
 #define CONFIG_ENV_IS_IN_MMC
-#define CONFIG_SYS_MMC_ENV_DEV          0
-#define CONFIG_SYS_MMC_ENV_PART         0
+#define CONFIG_SYS_MMC_ENV_DEV          1
+#define CONFIG_SYS_MMC_ENV_PART         1
 #define CONFIG_SYS_REDUNDAND_ENVIRONMENT
 
 #ifndef CONFIG_SPL_BUILD
-/* Enable gpt partition table */
-#undef PARTS_DEFAULT
-#define PARTS_DEFAULT \
-	"uuid_disk=${uuid_gpt_disk};" \
-	"name=loader_a,start=4M,size=4M,uuid=${uuid_gpt_loader};" \
-	"name=loader_b,size=4M,uuid=${uuid_gpt_reserved};" \
-	"name=trust_a,size=4M,uuid=${uuid_gpt_reserved};" \
-	"name=trust_b,size=4M,uuid=${uuid_gpt_reserved};" \
-	"name=misc,size=4M,uuid=${uuid_gpt_misc};" \
-	"name=metadata,size=16M,uuid=${uuid_gpt_metadata};" \
-	"name=boot_a,size=32M,uuid=${uuid_gpt_boot_a};" \
-	"name=boot_b,size=32M,uuid=${uuid_gpt_boot_b};" \
-	"name=system_a,size=818M,uuid=${uuid_gpt_system_a};" \
-	"name=system_b,size=818M,uuid=${uuid_gpt_system_b};" \
-	"name=vendor_a,size=50M,uuid=${uuid_gpt_vendor_a};" \
-	"name=vendor_b,size=50M,uuid=${uuid_gpt_vendor_b};" \
-	"name=cache,size=100M,uuid=${uuid_gpt_cache};" \
-	"name=persist,size=4M,uuid=${uuid_gpt_persist};" \
-	"name=userdata,size=-,uuid=${uuid_gpt_userdata};\0" \
+
 
 #define CONFIG_PREBOOT
 
 #define CONFIG_SYS_BOOT_RAMDISK_HIGH
 
-#undef CONFIG_BOOTCOMMAND
-#define CONFIG_BOOTCOMMAND \
-	"mmc read 0x61000000 0x8000 0x5000;" \
-	"bootm 0x61000000" \
 
 /* Enable atags */
 #define CONFIG_SYS_BOOTPARAMS_LEN	(64*1024)
@@ -54,6 +32,42 @@
 #define CONFIG_SETUP_MEMORY_TAGS
 #define CONFIG_CMDLINE_TAG
 
-#endif
 
+#include <config_distro_bootcmd.h>
+
+#undef CONFIG_EXTRA_ENV_SETTINGS
+#define CONFIG_EXTRA_ENV_SETTINGS \
+	"fdt_high=0x7fffffff\0" \
+	"partitions=" PARTS_DEFAULT \
+	ENV_MEM_LAYOUT_SETTINGS \
+	"earlycon=uart8250,mmio32,0x11030000\0" \
+	"console=uart8250,mmio32,0x11030000,115200n8 rw\0" \
+	"rootfstype=ext4\0" \
+	"init=/sbin/init\0" \
+	"respeaker_boot=" \
+		"if mmc dev 1 ; then " \
+			"mmcdev=1; " \
+		"else " \
+			"mmcdev=0; " \
+		"fi; " \
+		"echo SD/MMC found on device ${mmcdev}; " \
+		"part uuid mmc ${mmcdev}:7 rootfsuuid; " \
+		"echo [DEBUG: fatload mmc 0:6 ${scriptaddr} uEnv.txt]; " \
+		"fatload mmc 0:6 ${scriptaddr} uEnv.txt; " \
+		"env import -t ${scriptaddr}; " \
+		"setenv bootargs console=${console} " \
+			"${cmdline} " \
+			"root=PARTUUID=${rootfsuuid} " \
+			"rootfstype=${rootfstype}; " \
+		"fatload mmc ${mmcdev}:6 ${kernel_addr_r} vmlinuz-${uname_r}; " \
+		"fatload mmc ${mmcdev}:6 ${fdt_addr_r} ${dtb}; " \
+		"fatload mmc ${mmcdev}:6 ${ramdisk_addr_r} initrd.img-${uname_r}; " \
+			"bootz ${kernel_addr_r} ${ramdisk_addr_r}:${filesize} ${fdt_addr_r}\0" \
+	BOOTENV
+
+
+#undef CONFIG_BOOTCOMMAND
+#define CONFIG_BOOTCOMMAND "run respeaker_boot"
+
+#endif
 #endif
-- 
2.7.4

