From d33d2b78eb2fbba1df7d64cbdca183324a33b38a Mon Sep 17 00:00:00 2001
From: Baozhu Zuo <zuobaozhu@gmail.com>
Date: Tue, 8 Aug 2017 14:54:01 +0800
Subject: [PATCH 10/11] scripts: allow some waining in high gcc version

---
 scripts/gcc-wrapper.py | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/scripts/gcc-wrapper.py b/scripts/gcc-wrapper.py
index 1e0c1e2..370c890 100755
--- a/scripts/gcc-wrapper.py
+++ b/scripts/gcc-wrapper.py
@@ -49,6 +49,14 @@ allowed_warnings = set([
     "memcontrol.c:5337", # mm/memcontrol.c:5337:12: warning: initialization from incompatible pointer type
     "atags_to_fdt.c:98", # arch/arm/boot/compressed/atags_to_fdt.c:98:1: warning: the frame size of 1032 bytes is larger than 1024 bytes
     "drm_edid.c:3506", # drivers/gpu/drm/drm_edid.c:3506:13: warning: 'cea_db_is_hdmi_forum_vsdb' defined but not used
+    "rfkill-bt.c:703", # net/rfkill/rfkill-bt.c:703:5: warning: this ‘if’ clause does not guard...
+    "siutils.c:516", # siutils.c:516:14: warning: self-comparison always evaluates to false
+    "rtw_security.c:1598", # rtw_security.c:1598:5: warning: this ‘for’ clause does not guard...  
+    "rtw_security.c:1984", # rtw_security.c:1984:5: warning: this ‘for’ clause does not guard...
+    "wl_iw.c:605", # wl_iw.c:605:2: warning: this ‘if’ clause does not guard...
+    "rtw_mlme_ext.c:12253",# rtw_mlme_ext.c:12253:7: warning: this ‘else’ clause does not guard...
+    "rtw_mlme_ext.c:12256", # rtw_mlme_ext.c:12256:7: warning: this ‘if’ clause does not guard...
+    "rtw_iol.c:261", #  warning: this ‘if’ clause does not guard...
  ])
 
 # Capture the name of the object file, can find it.
-- 
2.7.4

