From aa324171b374c6250d884a9d45b9dd0f3b69229c Mon Sep 17 00:00:00 2001
From: Baozhu Zuo <zuobaozhu@gmail.com>
Date: Fri, 22 Sep 2017 17:33:13 +0800
Subject: [PATCH 2/5] sound: soc: open and close mclk in i2s_trigger

---
 sound/soc/rockchip/rockchip_i2s.c | 17 ++++++++++++-----
 1 file changed, 12 insertions(+), 5 deletions(-)

diff --git a/sound/soc/rockchip/rockchip_i2s.c b/sound/soc/rockchip/rockchip_i2s.c
index 0923a86..4a38262 100644
--- a/sound/soc/rockchip/rockchip_i2s.c
+++ b/sound/soc/rockchip/rockchip_i2s.c
@@ -395,17 +395,23 @@ static int rockchip_i2s_trigger(struct snd_pcm_substream *substream,
 	case SNDRV_PCM_TRIGGER_START:
 	case SNDRV_PCM_TRIGGER_RESUME:
 	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
-		if (substream->stream == SNDRV_PCM_STREAM_CAPTURE)
+		if (substream->stream == SNDRV_PCM_STREAM_CAPTURE){
 			rockchip_snd_rxctrl(i2s, 1);
-		else
+			ret = clk_prepare_enable(i2s->mclk);
+			if (ret) {
+				dev_err(i2s->dev, "mclk enable failed %d\n", ret);
+				return ret;
+			}
+		} else
 			rockchip_snd_txctrl(i2s, 1);
 		break;
 	case SNDRV_PCM_TRIGGER_SUSPEND:
 	case SNDRV_PCM_TRIGGER_STOP:
 	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
-		if (substream->stream == SNDRV_PCM_STREAM_CAPTURE)
+		if (substream->stream == SNDRV_PCM_STREAM_CAPTURE){
 			rockchip_snd_rxctrl(i2s, 0);
-		else
+			clk_disable_unprepare(i2s->mclk);
+		} else
 			rockchip_snd_txctrl(i2s, 0);
 		break;
 	default:
@@ -614,8 +620,9 @@ static int rockchip_i2s_probe(struct platform_device *pdev)
 		return PTR_ERR(i2s->mclk);
 	}
 
-	clk_prepare_enable(i2s->mclk);
+	//clk_prepare_enable(i2s->mclk);
 	clk_set_rate(i2s->mclk, 11289600);
+	clk_disable_unprepare(i2s->mclk);
 
 	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
 	regs = devm_ioremap_resource(&pdev->dev, res);
-- 
2.7.4

